`lib.fileset.fromSource`: init

infinisil opened the pull request:
## Description of changes

Another split off from https://github.com/NixOS/nixpkgs/pull/222981, motivated by @alyssais's comment [here](https://github.com/NixOS/nixpkgs/pull/234297#issuecomment-1767112029). This introduces one new function:
- `lib.fileset.fromSource`: Create a file set from a filtered local source as produced by the `lib.sources` functions.
  ```nix
  lib.fileset.toSource {
    root = ./.;
    fileset = lib.fileset.intersection
      (lib.fileset.fromSource (lib.sources.cleanSource ./.))
      (lib.fileset.unions [
        ./Makefile
        ./src
      ]);
  }
  ```

This is mainly useful for migration from `lib.sources`-based filtering. In the future we should have a convenience function in `lib.fileset` that could take over.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done

- [x] Tests
- [x] Clean code
- [x] Documentation

infinisil committed:
lib.fileset.fromSource: init

infinisil commented:
I'm wondering if this will still be necessary with https://github.com/NixOS/nixpkgs/pull/264891 :thinking:

alyssais commented:
Definitely is IMO. Not everything is or should be a git repository.

alyssais commented:
Another reason I can't switch from sources to filesets yet:

I want to be able to do, essentially: `src = fileset.difference ./. ./generated.o`, where generated.o is some generated file that may or may not exist in my development tree, but that shouldn't be input to the derivation.  This is not currently possible, because filesets require all files to exist.  So there needs to be a way to filter a fileset that doesn't require the files in the negative set to exist.

infinisil commented:
It is possible to do this already with

```nix
difference ./.
  (unions
    (lib.optional (builtins.pathExists ./generated.o) ./generated.o)
  )
```

But I think a convenient shorthand for this would be appropriate, I also had that in https://github.com/NixOS/nixpkgs/pull/222981, so I just created https://github.com/NixOS/nixpkgs/pull/265964 which adds a `lib.fileset.optional`!

fricklerhandwerk commented:
I'm worried the interface is growing faster than adoption of the library warrants. This issue could be stop-gapped with an example in the documentation, because the use case can be implemented with what's already there.

If it turns out to be a heavily used pattern one can still add a convenience layer.

fricklerhandwerk commented on lib/fileset/default.nix:
```diff
@@ -187,6 +188,24 @@ If a directory does not recursively contain any file, it is omitted from the sto
+    if ! source._isLibCleanSourceWith or false || ! source ? origSrc || ! source ? filter then
+      throw "lib.fileset.fromSource: Expected the argument to be a value produced from `lib.sources`, but got a ${typeOf source} instead."
+    else if ! isPath source.origSrc then
+      throw "lib.fileset.fromSource: Expected the argument to be source-like value of a local path."
```
What's a source-like value? What's a source?

roberth replied:
"Very" good question unfortunately. Basically a path value or the "source attrset" returned by the `lib.sources` functions. Can we write introductions for sublibraries yet? It'd be great to have a sensible place to document the unique types of each sublibrary.

infinisil replied:
> Can we write introductions for sublibraries yet? 

[Almost](https://github.com/nix-community/nixdoc/pull/70)

fricklerhandwerk reviewed (approved):
That's a really valuable addition to the library, as you said because it would ease migration a lot. We may want to note that in the docstring, along the lines of

> use this to gradually migrate builds relying on `lib.sources` to `lib.fileset`

but that's no reason to block this PR.

PS: Not sure why it's still in draft mode, looks ready to me.

fricklerhandwerk commented on lib/fileset/default.nix:
```diff
@@ -187,6 +188,24 @@ If a directory does not recursively contain any file, it is omitted from the sto
+    fromSource :: SourceLike -> FileSet
+
+  Example:
+    fromSource (lib.sources.cleanSource ./.)
```
The example in the PR description is also relevant, because it shows a bit of integration.

infinisil replied:
Added some more practical examples now

roberth reviewed (commented):
Remember to test that directories for which the predicate is `false` are not read. This is part of the (implicit) contract.
Other than that, I think the tests will "write themselves" if that makes sense.

infinisil commented:
> PS: Not sure why it's still in draft mode, looks ready to me.

See the checklist in the PR description :slightly_smiling_face:. Generally I recommend not taking close looks at the code of draft PRs. If I request you for a review it's because I'd like feedback to the approach only. I'll mark PRs as ready to review when I think their implementation is ready to review.

The review comments you both left are very helpful though, will incorporate!

infinisil committed:
lib.fileset: Split out internal test helper

infinisil commented:
> Other than that, I think the tests will "write themselves" if that makes sense.

It didn't make sense to me, but I wrote tests anyways and pushed them now :)

Still need to clean up the code and docs a bit, including addressing the review comments. I'll mark it as ready when everything is addressed.

infinisil commented on lib/fileset/internal.nix:
```diff
@@ -461,6 +461,44 @@ rec {
 
+  # Turn a builtins.filterSource-based source filter on a root path into a file set containing only files included by the filter
+  # Type: Path -> (String -> String -> Bool) -> fileset
+  _fromSource = root: filter:
```
`predicate` makes it a bit harder to understand, because it's really the `filter` argument to `builtins.path`/`builtins.filterSource`.

I now changed this to `sourceFilter` though, which both gets across that it's the source filter, and that it's not the same as `lib.filter`.

roberth replied:
That's an unfortunate name to have to inherit, but you've found a good solution.

infinisil commented on lib/fileset/default.nix:
```diff
@@ -187,6 +188,24 @@ If a directory does not recursively contain any file, it is omitted from the sto
+    fromSource (lib.sources.cleanSource ./.)
+  */
+  fromSource = source:
+    if ! source._isLibCleanSourceWith or false || ! source ? origSrc || ! source ? filter then
```
Done now!

infinisil committed:
lib.fileset.toSource: Mention fromSource in errors

roberth commented on lib/fileset/default.nix:
```diff
@@ -151,7 +153,12 @@ If a directory does not recursively contain any file, it is omitted from the sto
     if ! isPath root then
-      if isStringLike root then
+      if root ? _isLibCleanSourceWith then
+        throw ''
+          lib.fileset.toSource: `root` is a `lib.sources`-based value, but it should be a path instead.
+              To use a `lib.sources`-based value, convert it to a file set using `lib.fileset.fromSource` and pass it as `fileset`.
+              Note that this only works for sources created from paths.''
```
Any reason not to just call it for them?

infinisil replied:
I'm not 100% convinced it's a good idea, just minor ideas like
- The operation this function does is somewhat expensive, so it being explicit makes people aware of that. This is different from the implicit path coercion, which is very cheap
- Do we want to deprecate this at some point? Probably not
- Source filters can be impure by changing behavior based on the project directory, does this impurity leak into file sets? Probably not a problem
- Is it a problem to have a lack of symmetry regarding `toSource`/`fromSource`? Probably not, these aren't symmetric, fromSource discard's the `root` and `toSource` has to add it
- Would it be a problem to have cases like `toSource { root = ./.; fileset = toSource { root = ./.; fileset = toSource { ...`? Probably also no

Furthermore this being an error for now means that we could still do this later.

roberth reviewed (commented):
One question; otherwise lgtm.
